src/action.py

```diff
@@ -1,30 +1,79 @@
-import os
+import re
+
+from sys import exit
+from os import listdir
+from os.path import join, dirname, relpath, isfile, isdir
 from models.config import Config
 from collections import defaultdict
 from itertools import chain
 
+#TODO DRY up the file leader - only define '/var/lib/adminware/tools' once
 class ClickGlob:
-    def __glob_configs(namespace):
-        parts = ['/var/lib/adminware/tools',
-                 namespace, '*/config.yaml']
-        paths = glob.glob(os.path.join(*parts))
-        return list(map(lambda x: Config(x), paths))
+    def __glob_dirs(namespace):
+        parts = ['/var/lib/adminware/tools', namespace, '*']
+        paths = glob.glob(join(*parts))
+        return paths
+
+    # decorator for __glob_dir to get all valid paths (ending in config.yaml) at once
+    def __glob_all(globbing_func):
+        def __wrapper(*args, **kwargs):
+            def __sub_wrapper(*args, **kwargs):
+                paths = globbing_func(*args)
+                for path in paths:
+                    if isfile('{}/config.yaml'.format(path)):
+                        kwargs['all_paths'] += [join(path, 'config.yaml')]
+                    else:
+                        namespace = path[len('/var/lib/adminware/tools/'):]
+                        __sub_wrapper(namespace, **kwargs)
+                return kwargs['all_paths']
+
+            kwargs['all_paths']=[]
+            all_paths = __sub_wrapper(*args, **kwargs)
+            return all_paths
+
+        return __wrapper
 
     def command(click_group, namespace):
         # command_func is either run_open or run_batch, what this is decorating
         def __command(command_func):
-            actions = list(map(lambda x: Action(x), ClickGlob.__glob_configs(namespace)))
-            for action in actions:
-                action.create(click_group, command_func)
+            __command_helper(click_group, namespace, command_func)
+
+        def __command_helper(cur_group, cur_namespace, command_func):
+            paths = ClickGlob.__glob_dirs(cur_namespace) # will generate a list of paths at level 'namespace'
+            for path in paths:
+                if isfile('{}/config.yaml'.format(path)):
+                    # if there exists a sibling dir to any config.yaml this is currently an error
+                    # TODO this will be removed with closure of issue #49 but more validation may be needed
+                    if any(map(lambda x: isdir(join(path, x)), listdir(path))):
+                        print('config.yaml file with directory as sibling detected at {}\n'.format(path)
+                                + 'This is invalid, please rectify it before continuing')
+                        exit(1)
+
+                    action = Action(Config('{}/config.yaml'.format(path)))
+                    action.create(cur_group, command_func)
+                else:
+                    regex_expr = r'\/var\/lib\/adminware\/tools\/' + re.escape(cur_namespace) + r'\/(.*$)'
```

Comment:
We shouldn't be hard coding `/var/lib/adminware/tools` into the regex like this. Instead this path needs to be extracted as a constant some where. And the called on here and there https://github.com/alces-software/adminware/blob/8856765fa5c0348868178ff46d471016123a6d6e/src/action.py#L16.

Reply:
Also can this entire section be extracted to a `helper` module? Maybe a public method on `ClickGlob`?